student_work/julesfowler_work/reduce_pluto.py

```diff
@@ -0,0 +1,222 @@
+        Whether or not to normalize the flat to 1.
+    writeout : str or None
+        Name of the fits file to write out, if at all.
+
+    Returns
+    -------
+    flat : np.array
+        Array of dark subtracted and time averaged flat data.
+    """
+    super_flat = weighted_median_combine(flat_path)
+    super_flat_subtracted = super_flat - dark_image
+
+    if norm_to_1:
+        sigma_max = np.max(sigma_clip(super_flat_subtracted, masked=False, axis=None))
+        flat = super_flat_subtracted/sigma_max
+    else:
+        flat = super_flat_subtracted
+
+    if writeout is not None:
+        hdu_list = fits.HDUList([fits.PrimaryHDU(flat)])
+        hdu_list.writeto(writeout, overwrite=True)
+
+    return flat
+
+
+def make_science_image(science_image_path, bias_image, dark_image, flat_image, writeout=None, plot=None):
+    """ Builds the final reduced science image. 
+
+    Parameters
+    ----------
+    science_image_path: str
+        Glob-able path to the science fits files.
+    bias_image : np.array
+        Array of super bias data.
+    dark_image : np.array
+        Array of super dark data.
+    flat_image : np.array
+        Dark subtracted and time averaged flat data.
+    writeout : str or None
+        Name of the fits file to write out, if at all.
+    plot : str or None
+        Name of the plot to write out, if at all.
+
+    Returns
+    -------
+    reduced_science : np.array
+        Array of final reduced science image.
+    """
+
+    super_science = median_combine(science_image_path)
+    super_science_subtracted = super_science - bias_image
+    
+    science_exp_time = fits.getheader(glob.glob(science_image_path)[0])['EXPTIME']
+    super_science_per_second = super_science_subtracted/science_exp_time
+
+    reduced_science = (super_science_per_second - dark_image)/flat_image
+
+    if writeout is not None:
+        hdu_list = fits.HDUList([fits.PrimaryHDU(reduced_science)])
+        hdu_list.writeto(writeout, overwrite=True)
+
+    if plot is not None:
+        plt.imshow(reduced_science,
+                vmin=np.median(reduced_science)-3*np.std(reduced_science),
+                vmax=np.median(reduced_science)+3*np.std(reduced_science),
+                cmap='bone')
+        plt.savefig(plot)
+
+    return reduced_science
+
+
+def median_combine(path):
+    """ Median combines a given set of images.
+    
+    Parameters
+    ----------
+    path : str
+        Glob-able path of files.
+
+    Returns
+    -------
+    median : np.array
+        Array of median combined images.
+    """
+
+    files = glob.glob(path)
+    size = np.shape(fits.getdata(files[0]))
+    
+    cube = np.zeros((size[0], size[1], len(files)))
+    for index, fits_file in enumerate(files):
+        cube[:, :, index] = fits.getdata(fits_file)
+
+    median = np.median(cube, axis=2)
+
+    return median
+
+
+def weighted_median_combine(path):
+    """ Median combines a given set of images and weights them by their exposure time.
```

Comment:
This is great!

However, what if you have sky flats where the sky brightness
is varying?  What would you want to weight by in that case?
And did you?  :)

Reply:
Great point -- I should normalize to 1, then median combine for these exposures.